Applications/Spire/Source/Ui/TextBox.cpp

```diff
@@ -1,155 +1,95 @@
+      QLineEdit:read-only:disabled {
         background-color: #00000000;
         border: none;
-      })").
-        arg(BACKGROUND_COLOR.name(QColor::HexArgb)).arg(scale_width(1)).
-        arg(BORDER_COLOR.name(QColor::HexArgb)).arg(padding.m_left_padding).
-        arg(padding.m_right_padding);
-  }
-
-  auto get_warning_style_sheet(const QColor& background_color,
-      const QColor& border_color) {
-    return QString(R"(
-      #TextBox[%1=true] {
-        background-color: %2;
-        border-color: %3;
-      })").
-        arg(WARNING_PROPERTY_NAME).arg(background_color.name(QColor::HexArgb)).
-        arg(border_color.name(QColor::HexArgb));
-  }
-
-  auto get_fade_out_step(int start, int end) {
-    return (end - start) / WARNING_FADE_OUT_TIME_LINE_FRAME;
-  }
-
-  auto get_color_step(const QColor& start_color, const QColor& end_color) {
-    return std::array{get_fade_out_step(start_color.red(), end_color.red()),
-      get_fade_out_step(start_color.green(), end_color.green()),
-      get_fade_out_step(start_color.blue(), end_color.blue())};
-  }
-
-  const auto BACKGROUND_COLOR_STEP = get_color_step(BACKGROUND_WARNING_COLOR,
-    BACKGROUND_COLOR);
-  const auto BORDER_COLOR_STEP = get_color_step(BORDER_WARNING_COLOR,
-    BORDER_COLOR);
-
-  auto get_fade_out_color(const QColor& color, const std::array<int, 3>& steps,
-      int frame) {
-    return QColor(color.red() + steps[0] * frame,
-      color.green() + steps[1] * frame, color.blue() + steps[2] * frame);
+      })").arg(scale_width(1)).arg(scale_width(12)).arg(scale_width(8));
   }
 }
 
 TextBox::TextBox(QWidget* parent)
   : TextBox("", parent) {}
 
-TextBox::TextBox(const QString& text, QWidget* parent)
-    : QLineEdit(text, parent),
-      m_text(text),
-      m_submitted_text(text),
-      m_padding({scale_width(8), scale_width(8)}) {
-  setObjectName("TextBox");
-  setFrame(false);
-  auto font = QFont("Roboto");
-  font.setPixelSize(scale_width(12));
-  font.setWeight(QFont::Normal);
-  setFont(font);
-  m_style_sheet = get_text_box_style_sheet(m_padding);
-  setStyleSheet(m_style_sheet);
-  m_warning_time_line.setDuration(WARNING_FADE_OUT_TIME_MS);
-  m_warning_time_line.setFrameRange(0, WARNING_FADE_OUT_TIME_LINE_FRAME);
-  m_warning_time_line.setEasingCurve(QEasingCurve::Linear);
-  connect(&m_warning_time_line, &QTimeLine::frameChanged, this,
-    &TextBox::on_warning_fade_out);
-  connect(&m_warning_time_line, &QTimeLine::finished, this,
-    &TextBox::on_warning_finished);
-  connect(this, &QLineEdit::editingFinished, this,
+TextBox::TextBox(const QString& current, QWidget* parent)
+    : QWidget(parent),
+      m_current(current) {
+  m_line_edit = new QLineEdit(m_current, this);
+  m_line_edit->setFrame(false);
+  m_line_edit->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins(0, 0, 0, 0);
+  layout->addWidget(m_line_edit);
+  setLayout(layout);
```

Comment:
You don't have to manually set the layout here, because constructing the layout with a parent automatically sets the layout as the parent's main layout: https://doc.qt.io/qt-5/qhboxlayout.html#QHBoxLayout-1